lib/posix/pthread.c

```diff
@@ -154,70 +178,54 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
 int pthread_create(pthread_t *newthread, const pthread_attr_t *_attr,
 		   void *(*threadroutine)(void *), void *arg)
 {
-	int rv;
+	bool found;
 	int32_t prio;
-	k_spinlock_key_t key;
 	uint32_t pthread_num;
-	k_spinlock_key_t cancel_key;
-	pthread_condattr_t cond_attr;
+	k_spinlock_key_t key;
 	struct posix_thread *thread;
 	const struct pthread_attr *attr = (const struct pthread_attr *)_attr;
 
-	/*
-	 * FIXME: Pthread attribute must be non-null and it provides stack
-	 * pointer and stack size. So even though POSIX 1003.1 spec accepts
-	 * attrib as NULL but zephyr needs it initialized with valid stack.
-	 */
-	if ((attr == NULL) || (attr->initialized == 0U)
-	    || (attr->stack == NULL) || (attr->stacksize == 0)) {
+	if (!pthread_attr_is_valid(attr)) {
 		return EINVAL;
 	}
 
-	key = k_spin_lock(&pthread_pool_lock);
-	for (pthread_num = 0;
-	    pthread_num < CONFIG_MAX_PTHREAD_COUNT; pthread_num++) {
+	for (found = false, pthread_num = 0; !found && pthread_num < CONFIG_MAX_PTHREAD_COUNT;
+	     pthread_num++) {
+
 		thread = &posix_thread_pool[pthread_num];
-		if (thread->state == PTHREAD_EXITED || thread->state == PTHREAD_TERMINATED) {
-			thread->state = PTHREAD_JOINABLE;
-			break;
-		}
-	}
-	k_spin_unlock(&pthread_pool_lock, key);
 
-	if (pthread_num >= CONFIG_MAX_PTHREAD_COUNT) {
-		return EAGAIN;
-	}
+		key = k_spin_lock(&thread->lock);
+		if (thread->state == PTHREAD_TERMINATED) {
 
-	rv = pthread_mutex_init(&thread->state_lock, NULL);
-	if (rv != 0) {
-		key = k_spin_lock(&pthread_pool_lock);
-		thread->state = PTHREAD_EXITED;
-		k_spin_unlock(&pthread_pool_lock, key);
-		return rv;
-	}
+			found = true;
+			thread->state = attr->detachstate;
 
-	prio = posix_to_zephyr_priority(attr->priority, attr->schedpolicy);
+			prio = posix_to_zephyr_priority(attr->priority, attr->schedpolicy);
 
-	cancel_key = k_spin_lock(&thread->cancel_lock);
-	thread->cancel_state = (1 << _PTHREAD_CANCEL_POS) & attr->flags;
-	thread->cancel_pending = 0;
-	k_spin_unlock(&thread->cancel_lock, cancel_key);
+			thread->cancel_state = BIT(_PTHREAD_CANCEL_POS) & attr->flags;
+			thread->cancel_pending = 0;
 
-	pthread_mutex_lock(&thread->state_lock);
-	thread->state = attr->detachstate;
-	pthread_mutex_unlock(&thread->state_lock);
+			sys_slist_init(&thread->key_list);
+		}
+		k_spin_unlock(&thread->lock, key);
 
-	pthread_cond_init(&thread->state_cond, &cond_attr);
-	sys_slist_init(&thread->key_list);
+		if (found) {
```

Comment:
this looks redundant with the loop condition?

Reply:
Fixed - thanks for spotting that.